File: packages/research_management/src/research_management/models/project.py
"""
Research Project models.
"""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class ProjectStatus(str, Enum):
    """Project lifecycle status."""

    PROPOSAL = "proposal"
    ACTIVE = "active"
    PAUSED = "paused"
    COMPLETED = "completed"
    ARCHIVED = "archived"


class HealthStatus(str, Enum):
    """Project health indicators."""

    ON_TRACK = "on_track"  # 🟢 Everything OK
    AT_RISK = "at_risk"  # 🟡 Minor delays or communication issues
    CRITICAL = "critical"  # 🔴 Significant delays or imminent abandonment


class ResearchProject(BaseModel):
    """Research project model."""

    project_id: str = Field(..., description="Unique project identifier")
    title: str = Field(..., min_length=1, max_length=200, description="Project title")
    description: str = Field(..., description="Project description")
    area: str = Field(
        ..., description="Research area (CS, ML, Networks, etc)", max_length=100
    )
    status: ProjectStatus = Field(
        default=ProjectStatus.PROPOSAL, description="Current project status"
    )
    health_status: HealthStatus = Field(
        default=HealthStatus.ON_TRACK, description="Project health indicator"
    )
    start_date: Optional[datetime] = Field(None, description="Project start date")
    expected_end_date: Optional[datetime] = Field(
        None, description="Expected completion date"
    )
    actual_end_date: Optional[datetime] = Field(
        None, description="Actual completion date"
    )
    created_at: datetime = Field(
        default_factory=datetime.utcnow, description="Creation timestamp"
    )
    updated_at: datetime = Field(
        default_factory=datetime.utcnow, description="Last update timestamp"
    )

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "project_id": "proj-123",
                "title": "ML for Image Classification",
                "description": "Research project on machine learning applications",
                "area": "Machine Learning",
                "status": "active",
                "health_status": "on_track",
                "start_date": "2025-01-15T00:00:00Z",
                "expected_end_date": "2025-06-30T00:00:00Z",
                "created_at": "2025-01-10T10:00:00Z",
                "updated_at": "2025-01-10T10:00:00Z",
            }
        }
    )


class ProjectCreate(BaseModel):
    """Model for creating a new project."""

    title: str = Field(..., min_length=1, max_length=200)
    description: str
    area: str = Field(..., max_length=100)
    start_date: Optional[datetime] = None
    expected_end_date: Optional[datetime] = None


class ProjectUpdate(BaseModel):
    """Model for updating a project."""

    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    area: Optional[str] = Field(None, max_length=100)
    status: Optional[ProjectStatus] = None
    health_status: Optional[HealthStatus] = None
    start_date: Optional[datetime] = None
    expected_end_date: Optional[datetime] = None
    actual_end_date: Optional[datetime] = None

File: packages_dashboard/research_dashboard/lib/src/models/project.dart
/// Research project model
class ResearchProject {
  final String projectId;
  final String title;
  final String description;
  final String area;
  final ProjectStatus status;
  final HealthStatus healthStatus;
  final DateTime? startDate;
  final DateTime? expectedEndDate;
  final DateTime? actualEndDate;
  final DateTime createdAt;
  final DateTime updatedAt;

  const ResearchProject({
    required this.projectId,
    required this.title,
    required this.description,
    required this.area,
    required this.status,
    required this.healthStatus,
    this.startDate,
    this.expectedEndDate,
    this.actualEndDate,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ResearchProject.fromJson(Map<String, dynamic> json) {
    return ResearchProject(
      projectId: json['project_id'] as String,
      title: json['title'] as String,
      description: json['description'] as String,
      area: json['area'] as String,
      status: ProjectStatus.fromString(json['status'] as String),
      healthStatus: HealthStatus.fromString(json['health_status'] as String),
      startDate: json['start_date'] != null
          ? DateTime.parse(json['start_date'] as String)
          : null,
      expectedEndDate: json['expected_end_date'] != null
          ? DateTime.parse(json['expected_end_date'] as String)
          : null,
      actualEndDate: json['actual_end_date'] != null
          ? DateTime.parse(json['actual_end_date'] as String)
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'project_id': projectId,
      'title': title,
      'description': description,
      'area': area,
      'status': status.value,
      'health_status': healthStatus.value,
      'start_date': startDate?.toIso8601String(),
      'expected_end_date': expectedEndDate?.toIso8601String(),
      'actual_end_date': actualEndDate?.toIso8601String(),
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}

/// Project status enum
enum ProjectStatus {
  proposal('proposal'),
  active('active'),
  paused('paused'),
  completed('completed'),
  archived('archived');

  final String value;
  const ProjectStatus(this.value);

  static ProjectStatus fromString(String value) {
    return ProjectStatus.values.firstWhere((e) => e.value == value);
  }

  String get displayName {
    switch (this) {
      case ProjectStatus.proposal:
        return 'Proposta';
      case ProjectStatus.active:
        return 'Ativo';
      case ProjectStatus.paused:
        return 'Pausado';
      case ProjectStatus.completed:
        return 'Concluído';
      case ProjectStatus.archived:
        return 'Arquivado';
    }
  }
}

/// Project health status enum
enum HealthStatus {
  onTrack('on_track'),
  atRisk('at_risk'),
  critical('critical');

  final String value;
  const HealthStatus(this.value);

  static HealthStatus fromString(String value) {
    return HealthStatus.values.firstWhere((e) => e.value == value);
  }

  String get displayName {
    switch (this) {
      case HealthStatus.onTrack:
        return '🟢 No Prazo';
      case HealthStatus.atRisk:
        return '🟡 Em Risco';
      case HealthStatus.critical:
        return '🔴 Crítico';
    }
  }

  String get emoji {
    switch (this) {
      case HealthStatus.onTrack:
        return '🟢';
      case HealthStatus.atRisk:
        return '🟡';
      case HealthStatus.critical:
        return '🔴';
    }
  }
}
